README.md
# vdinar-rpc
[![npm version][npm-image]][npm-url]

A JavaScript RPC client for vDinar to execute administrative tasks, queries about the network and blockchain.

## Installation

Install the package via `npm`:

```sh
npm install @vdinar/vdinar-rpc --save
```

## Usage
### Client(...args)
#### Arguments
1. `[agentOptions]` _(Object)_: Optional `agent` [options](https://github.com/request/request#using-optionsagentoptions) to configure SSL/TLS.
2. `[headers=false]` _(boolean)_: Whether to return the response headers.
3. `[host=localhost]` _(string)_: The host to connect to.
4. `[logger=debugnyan('vdinar-rpc')]` _(Function)_: Custom logger (by default, `debugnyan`).
5. `[network=mainnet]` _(string)_: The network
6. `[password]` _(string)_: The RPC server user password.
7. `[port=[network]]` _(string)_: The RPC server port.
8. `[ssl]` _(boolean|Object)_: Whether to use SSL/TLS with strict checking (_boolean_) or an expanded config (_Object_).
9. `[ssl.enabled]` _(boolean)_: Whether to use SSL/TLS.
10. `[ssl.strict]` _(boolean)_: Whether to do strict SSL/TLS checking (certificate must match host).
11. `[timeout=30000]` _(number)_: How long until the request times out (ms).
12. `[username]` _(number)_: The RPC server user name.
13. `[version]` _(string)_: Which version to check methods for ([read more](#versionchecking)).

### Examples
#### Using network mode
The `network` will automatically determine the port to connect to, just like the `vdinard` command.

```js
const Client = require('@vdinar/vdinar-rpc');
const client = new Client({ network: 'mainnet' });
```

##### Setting a custom port

```js
const client = new Client({ port: 9334 });
```

#### Connecting to an SSL/TLS server with strict checking enabled
By default, when `ssl` is enabled, strict checking is implicitly enabled.

```js
const fs = require('fs');
const client = new Client({
  agentOptions: {
    ca: fs.readFileSync('/etc/ssl/vdinard/cert.pem')
  },
  ssl: true
});
```

#### Connecting to an SSL/TLS server without strict checking enabled

```js
const client = new Client({
  ssl: {
    enabled: true,
    strict: false
  }
});
```

#### Using promises to process the response

```js
client.getInfo().then((help) => console.log(help));
```

#### Using callbacks to process the response

```js
client.getInfo((error, help) => console.log(help));
```

#### Returning headers in the response
For compatibility with other vDinar RPC clients.

```js
const client = new Client({ headers: true });

// Promise style with headers enabled:
client.getInfo().then(([body, headers]) => console.log(body, headers));

// Await style based on promises with headers enabled:
const [body, headers] = await client.getInfo();
```

## Named parameters

It is possible to send commands via the JSON-RPC interface using named parameters instead of positional ones. This comes with the advantage of making the order of arguments irrelevant. It also helps improving the readability of certain function calls when leaving out arguments for their default value.

You **must** provide a version in the client arguments to enable named parameters.

```js
const client = new Client({ version: '2.1.0' });
```
For instance, take the `getBalance()` call written using positional arguments:
```js
const balance = await new Client().getBalance('*', 0);
```

It is functionally equivalent to using the named arguments `account` and `minconf`:

```js
const balance = await new Client({ version: '2.1.0' }).getBalance({
  account: '*',
  minconf: 0
});
```

This feature is available to all JSON-RPC methods that accept arguments.

### Floating point number precision in JavaScript

Due to [JavaScript's limited floating point precision](http://floating-point-gui.de/), all big numbers (numbers with more than 15 significant digits) are returned as strings to prevent precision loss. This includes both the RPC and REST APIs.

### Version Checking
By default, all methods are exposed on the client independently of the version it is connecting to. This is the most flexible option as defining methods for unavailable RPC calls does not cause any harm and the library is capable of handling a `Method not found` response error correctly.

```js
const client = new Client();

client.command('foobar');
// => RpcError: -32601 Method not found
```

However, if you prefer to be on the safe side, you can enable strict version checking. This will validate all method calls before executing the actual RPC request:

```js
const client = new Client({ version: '2.0.0' });

client.getHashesPerSec();
// => Method "getnetworkinfo" is not supported by version "2.0.0"
```

If you want to enable strict version checking for the bleeding edge version, you may set a very high version number to exclude recently deprecated calls:

```js
const client = new Client({ version: `${Number.MAX_SAFE_INTEGER}.0.0` });

client.getWork();
// => Throws 'Method "getwork" is not supported by version "9007199254740991.0.0"'.
```

To avoid potential issues with prototype references, all methods are still enumerable on the library client prototype.

### RPC
Start `vdinard` with the RPC server enabled and optionally configure a username and password:

```sh
docker run --rm -it AndreaDejanGrande/vdinar-rpc -printtoconsole -rpcuser=foo -rpcpassword=bar -server
```

These configuration values may also be set on the `vdinar.conf` file of your platform installation.

By default, port `9333` is used to listen for requests in `mainnet` mode, or `19333` in `testnet`. Use the `network` property to initialize the client on the desired mode and automatically set the respective default port. You can optionally set a custom port of your choice too.

The RPC services binds to the localhost loopback network interface, so use `rpcbind` to change where to bind to and `rpcallowip` to whitelist source IP access.

#### Methods
All RPC [methods](src/methods.js) are exposed on the client interface as a camelcase'd version of those available on `vdinard` (see examples below).

##### Examples

```js
client.createRawTransaction([{ txid: '1eb590cd06127f78bf38ab4140c4cdce56ad9eb8886999eb898ddf4d3b28a91d', vout: 0 }], { 'mgnucj8nYqdrPFh2JfZSB1NmUThUGnmsqe': 0.13 });
client.sendMany('test1', { mjSk1Ny9spzU2fouzYgLqGUD8U41iR35QN: 0.1, mgnucj8nYqdrPFh2JfZSB1NmUThUGnmsqe: 0.2 }, 6, 'Example Transaction');
client.sendToAddress('mmXgiR6KAhZCyQ8ndr2BCfEq1wNG2UnyG6', 0.1,  'sendtoaddress example', 'Nemo From Example.com');
```

#### Batch requests
Batch requests are support by passing an array to the `command` method with a `method` and optionally, `parameters`. The return value will be an array with all the responses.

```js
const batch = [
  { method: 'getnewaddress', parameters: [] },
  { method: 'getnewaddress', parameters: [] }
]

new Client().command(batch).then((responses) => console.log(responses)));

// Or, using ES2015 destructuring.
new Client().command(batch).then(([firstAddress, secondAddress]) => console.log(firstAddress, secondAddress)));
```

Note that batched requests will only throw an error if the batch request itself cannot be processed. However, each individual response may contain an error akin to an individual request.

```js
const batch = [
  { method: 'foobar', params: [] },
  { method: 'getnewaddress', params: [] }
]

new Client().command(batch).then(([address, error]) => console.log(address, error)));
// => `mkteeBFmGkraJaWN5WzqHCjmbQWVrPo5X3, { [RpcError: Method not found] message: 'Method not found', name: 'RpcError', code: -32601 }`.
```


```js
client.getBlockchainInformation([callback]);
```

### SSL
This client supports SSL out of the box. Simply pass the SSL public certificate to the client and optionally disable strict SSL checking which will bypass SSL validation (the connection is still encrypted but the server it is connecting to may not be trusted). This is, of course, discouraged unless for testing purposes when using something like self-signed certificates.

#### Generating a self-signed certificates for testing purposes
Please note that the following procedure should only be used for testing purposes.

Generate an self-signed certificate together with an unprotected private key:

```sh
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 3650 -nodes
```

## Logging

By default, all requests made with `vdinar-rpc` are logged using [uphold/debugnyan](https://github.com/uphold/debugnyan) with `vdinar-rpc` as the logging namespace.

Please note that all sensitive data is obfuscated before calling the logger.

### Custom logger

A custom logger can be passed via the `logger` option and it should implement [bunyan's log levels](https://github.com/trentm/node-bunyan#levels).

## Tests
Currently the test suite is tailored for Docker (including `docker-compose`) due to the multitude of different `vdinard` configurations that are required in order to get the test suite passing.

To test using a local installation of `node.js` but with dependencies (e.g. `vdinard`) running inside Docker:

```sh
npm run dependencies
npm test
```

To test using Docker exclusively (similarly to what is done in Travis CI):

```sh
npm run testdocker
```

## Release

```sh
npm version [<newversion> | major | minor | patch] -m "Release %s"
```

## License
MIT

[npm-image]: https://img.shields.io/npm/v/@vdinar/vdinar-rpc.svg
[npm-url]: https://npmjs.org/package/@vdinar/vdinar-rpc


CHANGELOG.md
# Changelog

## [0.1.0](https://github.com/AndreaDejanGrande/vdinar-rpc/tree/0.1.0) (2019-04-29)

- Removed multiwallet and REST functions as they are not supported in vDinar
- Added vDinar RPC methods
- Updated main files and package details
- Tests made compatible for vDinar


test/client_test.js

/**
 * Module dependencies.
 */

import Client from '../src/index';
import RpcError from '../src/errors/rpc-error';
import _ from 'lodash';
import config from './config';
import fs from 'fs';
import methods from '../src/methods';
import parse from './utils/help-parser-util';
import path from 'path';
import should from 'should';

/**
 * Test `Client`.
 */

describe('Client', () => {
  before(async () => {
    const client = new Client(config.vdinar);
  });

  describe('constructor', () => {
    it('should throw an error if network is invalid', () => {
      try {
        new Client(_.defaults({ network: 'foo' }, config.vdinar)); // eslint-disable-line no-new

        should.fail();
      } catch (error) {
        error.should.be.an.instanceOf(Error);
        error.message.should.equal('Invalid network name "foo"');
      }
    });

    it('should not return headers by default', () => {
      new Client().headers.should.be.false();
    });

    it('should have default host set to `localhost`', () => {
      new Client().host.should.equal('localhost');
    });

    it('should not have a password set by default', () => {
      should.not.exist(new Client().password);
    });

    it('should have default port set to `mainnet`\'s one', () => {
      new Client().port.should.equal(9333);
    });

    it('should set default to port `9333` if network is `mainnet`', () => {
      new Client({ network: 'mainnet' }).port.should.equal(9333);
    });

    it('should set default to port `19333` if network is `testnet`', () => {
      new Client({ network: 'testnet' }).port.should.equal(19333);
    });

    it('should not have ssl enabled by default', () => {
      new Client().ssl.enabled.should.equal(false);
      new Client().ssl.strict.should.equal(false);
    });

    it('should have default timeout of 30000ms', () => {
      new Client().timeout.should.equal(30000);
    });

    it('should not have username/password authentication enabled by default', () => {
      should.not.exist(new Client().auth);
    });

    it('should have all the methods listed by `help`', async () => {
      const help = await new Client(config.vdinar).help();

      _.difference(parse(help), _.invokeMap(Object.keys(methods), String.prototype.toLowerCase)).should.be.empty();
    });

    it('should accept valid versions', async () => {
      await new Client(_.defaults({ version: '2.1.0.1' }, config.vdinar)).getInfo();
      await new Client(_.defaults({ version: '2.1.0' }, config.vdinar)).getInfo();
    });
  });

  describe('connections', () => {
    describe('general', () => {
      it('should throw an error if timeout is reached', async () => {
        try {
          await new Client(_.defaults({ timeout: 0.1 }, config.vdinar)).listAccounts();

          should.fail();
        } catch (e) {
          e.should.be.an.instanceOf(Error);
          e.code.should.match(/(ETIMEDOUT|ESOCKETTIMEDOUT)/);
        }
      });

      it('should throw an error if version is invalid', async () => {
        try {
          await new Client({ version: '0.12' }).getHashesPerSec();

          should.fail();
        } catch (e) {
          e.should.be.an.instanceOf(Error);
          e.message.should.equal('Invalid Version "0.12"');
        }
      });

      it('should throw an error if a connection cannot be established', async () => {
        try {
          await new Client(_.defaults({ port: 9897 }, config.vdinar)).getDifficulty();

          should.fail();
        } catch (e) {
          e.should.be.an.instanceOf(Error);
          e.message.should.match(/connect ECONNREFUSED/);
          e.code.should.equal('ECONNREFUSED');
        }
      });
    });

    describe('ssl', () => {
      it('should use `ssl.strict` by default when `ssl` is enabled', () => {
        const sslClient = new Client(_.defaults({ host: config.vdinarSsl.host, port: config.vdinarSsl.port, ssl: true }, config.vdinar));

        sslClient.ssl.strict.should.be.true();
      });

      it('should throw an error if certificate is self signed by default', async () => {
        const sslClient = new Client(_.defaults({ host: config.vdinarSsl.host, port: config.vdinarSsl.port, ssl: true }, config.vdinar));

        sslClient.ssl.strict.should.be.true();

        try {
          await sslClient.getInfo();
        } catch (e) {
          e.should.be.an.instanceOf(Error);
          e.code.should.equal('DEPTH_ZERO_SELF_SIGNED_CERT');
          e.message.should.equal('self signed certificate');
        }
      });

      it('should establish a connection if certificate is self signed but `ca` agent option is passed', async () => {
        const sslClient = new Client(_.defaults({
          agentOptions: {
            /* eslint-disable no-sync */
            ca: fs.readFileSync(path.join(__dirname, '/config/ssl/cert.pem')),
            checkServerIdentity() {
              // Skip server identity checks otherwise the certificate would be immediately rejected
              // as connecting to an IP not listed in the `altname` fails.
              return;
            }
          },
          host: config.vdinarSsl.host,
          port: config.vdinarSsl.port,
          ssl: true
        }, config.vdinar));

        const info = await sslClient.getInfo();

        info.should.not.be.empty();
      });

      it('should establish a connection if certificate is self signed but `ssl.strict` is disabled', async () => {
        const sslClient = new Client(_.defaults({ host: config.vdinarSsl.host, port: config.vdinarSsl.port, ssl: { enabled: true, strict: false } }, config.vdinar));
        const info = await sslClient.getInfo();

        info.should.not.be.empty();
      });
    });

    describe('authentication', () => {
      it('should throw an error if credentials are invalid', async () => {
        try {
          await new Client(_.defaults({ password: 'biz', username: 'foowrong' }, config.vdinar)).getDifficulty();
        } catch (e) {
          e.should.be.an.instanceOf(RpcError);
          e.message.should.equal('Unauthorized');
          e.body.should.equal('');
          e.code.should.equal(401);
        }
      });

      it('should support username only authentication', async () => {
        const difficulty = await new Client(config.vdinarUsernameOnly).getDifficulty();

        difficulty.should.equal(0);
      });
    });
  });

  describe('callbacks', () => {
    it('should support callbacks', done => {
      new Client(config.vdinar).help((err, help) => {
        should.not.exist(err);

        help.should.not.be.empty();

        done();
      });
    });
  });
});


test/parser_test.js

/**
 * Module dependencies.
 */

import { defaults } from 'lodash';
import Client from '../src/index';
import RpcError from '../src/errors/rpc-error';
import config from './config';
import nock from 'nock';
import should from 'should';

/**
 * Test `Parser`.
 */

afterEach(() => {
  if (nock.pendingMocks().length) {
    throw new Error('Unexpected pending mocks');
  }

  nock.cleanAll();
});

describe('Parser', () => {
  it('should throw an error with a generic message if one is not returned on the response', async () => {
    nock(`http://${config.vdinar.host}:${config.vdinar.port}/`)
      .post('/')
      .reply(200, '{ "result": null, "error": { "code": -32601 }, "id": "69837016239933"}');

    try {
      await new Client(config.vdinar).command('foobar');

      should.fail();
    } catch (e) {
      e.should.be.an.instanceOf(RpcError);
      e.message.should.equal('An error occurred while processing the RPC call to vdinard');
      e.code.should.equal(-32601);
    }
  });

  it('should throw an error if the response does not include a `result`', async () => {
    nock(`http://${config.vdinar.host}:${config.vdinar.port}/`)
      .post('/')
      .reply(200, '{ "error": null, "id": "69837016239933"}');

    try {
      await new Client(config.vdinar).command('foobar2');

      should.fail();
    } catch (e) {
      e.should.be.an.instanceOf(RpcError);
      e.message.should.equal('Missing `result` on the RPC call result');
      e.code.should.equal(-32700);
    }
  });

  describe('headers', () => {
    it('should return the response headers if `headers` is enabled', async () => {
      const [info, headers] = await new Client(defaults({ headers: true }, config.vdinar)).getInfo();

      info.should.be.an.Object();
      headers.should.have.keys('date', 'connection', 'content-length', 'content-type');
    });

    it('should return the response headers if `headers` is enabled using callbacks', done => {
      new Client(defaults({ headers: true }, config.vdinar)).getInfo((err, [info, headers]) => {
        should.not.exist(err);

        info.should.be.an.Object();

        headers.should.have.keys('date', 'connection', 'content-length', 'content-type');

        done();
      });
    });

    it('should return the response headers if `headers` is enabled and batching is used', async () => {
      const batch = [
        { method: 'getbalance' },
        { method: 'getbalance' }
      ];
      const [addresses, headers] = await new Client(defaults({ headers: true }, config.vdinar)).command(batch);

      addresses.should.have.length(batch.length);
      headers.should.have.keys('date', 'connection', 'content-length', 'content-type');
    });

    it('should return the response headers if `headers` is enabled and batching is used with callbacks', done => {
      const batch = [
        { method: 'getbalance' },
        { method: 'getbalance' }
      ];

      new Client(defaults({ headers: true }, config.vdinar)).command(batch, (err, [addresses, headers]) => {
        should.not.exist(err);

        addresses.should.have.length(batch.length);

        headers.should.have.keys('date', 'connection', 'content-length', 'content-type');

        done();
      });
    });
  });
});


test/wallet_test.js

/**
 * Module dependencies.
 */

import { defaults } from 'lodash';
import Client from '../src/index';
import RpcError from '../src/errors/rpc-error';
import config from './config';

/**
 * Test instance.
 */

const client = new Client(config.vdinar);

/**
 * Test `Client`.
 */

describe('Single Wallet', () => {
  describe('node-level requests', () => {
    describe('getDifficulty()', () => {
      it('should return the proof-of-work difficulty', async () => {
        const difficulty = await client.getDifficulty();

        difficulty.should.be.a.String();
      });
    });
  });

  describe('wallet-level requests', () => {
    describe('getAccountAddress()', () => {
      it('should retrieve an account address', async () => {
        const address = await client.getAccountAddress('test');
        const account = await client.getAccount(address);

        account.should.equal('test');
      });
    });

    describe('getBalance()', () => {
      it('should return the total server\'s balance', async () => {
        const balance = await client.getBalance();

        balance.should.be.aboveOrEqual(0);
      });

      it('should support named parameters', async () => {
        const client = new Client(defaults({ version: '2.1.0' }, config.vdinar));

        const mainWalletBalance = await client.getBalance({ account: '*', minconf: 0 });
        const mainWalletBalanceWithoutParameters = await client.getBalance('*', 0);
        const testWalletBalance = await client.getBalance({ account: 'test', minconf: 0 });

        mainWalletBalance.should.not.equal(testWalletBalance);
        mainWalletBalanceWithoutParameters.should.equal(mainWalletBalance);
      });
    });

    describe('getNewAddress()', () => {
      it('should return a new vDinar address', async () => {
        await client.getNewAddress('test');

        const addresses = await client.getAddressesByAccount('test');

        addresses.length.should.be.above(1);
      });
    });

    describe('listTransactions()', () => {
      it('should return the most recent list of transactions from all accounts using specific count', async () => {
        const address = await client.getNewAddress('test');

        // Generate 5 transactions.
        for (let i = 0; i < 5; i++) {
          await client.sendToAddress(address, 0.1);
        }

        const transactions = await client.listTransactions('test', 5);

        transactions.should.be.an.Array();
        transactions.length.should.be.greaterThanOrEqual(5);
      });

      it('should return the most recent list of transactions from all accounts using default count', async () => {
        const transactions = await client.listTransactions('test');

        transactions.should.be.an.Array();
        transactions.should.matchEach(value => {
          // Only a small subset of transaction properties are being asserted here to make
          // sure we've received a transaction and not an empty object instead.
          value.should.have.keys(
            'account',
            'address',
            'category',
            'amount',
            'confirmations',
            'time',
            'txid'
          );
        });
      });

      it('should support named parameters', async () => {
        const address = await client.getNewAddress('test');

        // Generate 5 transactions.
        for (let i = 0; i < 5; i++) {
          await client.sendToAddress(address, 0.1);
        }

        let transactions = await new Client(defaults({ version: '2.1.0' }, config.vdinar)).listTransactions({ account: 'test' });

        transactions.should.be.an.Array();
        transactions.length.should.be.greaterThanOrEqual(5);

        // Make sure `count` is read correctly.
        transactions = await new Client(defaults({ version: '2.1.0' }, config.vdinar)).listTransactions({ account: 'test', count: 1 });

        transactions.should.be.an.Array();
        transactions.should.have.length(1);
      });
    });
  });

  describe('batched requests', () => {
    it('should support request parameters in batched requests', async () => {
      const batch = [{ method: 'getnewaddress' }, { method: 'validateaddress', parameters: ['DG1KpSsSXd3uitgwHaA1i6T1Bj1hWEwAxB'] }];

      const [newAddress, addressValidation] = await client.command(batch);

      addressValidation.should.have.properties('address', 'ismine', 'isvalid', 'scriptPubKey');
      newAddress.should.be.a.String();
    });

    it('should return an error if one of the request fails', async () => {
      const batch = [{ method: 'validateaddress' }, { method: 'validateaddress', parameters: ['DG1KpSsSXd3uitgwHaA1i6T1Bj1hWEwAxB'] }];

      const [validateAddressError, validateAddress] = await client.command(batch);

      validateAddress.should.have.properties('address', 'ismine', 'isvalid', 'scriptPubKey');
      validateAddressError.should.be.an.instanceOf(RpcError);
      validateAddressError.code.should.equal(-1);
    });
  });
});


src/parser.js

/**
 * Module dependencies.
 */

import JSONBigInt from 'json-bigint';
import RpcError from './errors/rpc-error';
import _ from 'lodash';

/**
 * JSONBigInt parser.
 */

const { parse } = JSONBigInt({ storeAsString: true, strict: true }); // eslint-disable-line new-cap

/**
 * Get RPC response body result.
 */

function getRpcResult(body, { headers = false, response } = {}) {
  if (body.error !== null) {
    throw new RpcError(
      _.get(body, 'error.code', -32603),
      _.get(body, 'error.message', 'An error occurred while processing the RPC call to vdinard')
    );
  }

  // Defensive measure. This should not happen on a RPC call.
  if (!_.has(body, 'result')) {
    throw new RpcError(-32700, 'Missing `result` on the RPC call result');
  }

  if (headers) {
    return [body.result, response.headers];
  }

  return body.result;
}

/**
 * Export Parser class.
 */

export default class Parser {
  constructor({ headers } = {}) {
    this.headers = headers;
  }

  /**
   * Parse rpc response.
   */

  rpc([response, body]) {
    // The RPC api returns a `text/html; charset=ISO-8859-1` encoded response with an empty string as the body
    // when an error occurs.
    if (typeof body === 'string' && response.headers['content-type'] !== 'application/json' && response.statusCode !== 200) {
      throw new RpcError(response.statusCode, response.statusMessage, { body });
    }

    // Parsing the body with custom parser to support BigNumbers.
    body = parse(body);

    if (!Array.isArray(body)) {
      return getRpcResult(body, { headers: this.headers, response });
    }

    // Batch response parsing where each response may or may not be successful.
    const batch = body.map(response => {
      try {
        return getRpcResult(response, { headers: false, response });
      } catch (e) {
        return e;
      }
    });

    if (this.headers) {
      return [batch, response.headers];
    }

    return batch;
  }
}


src/index.js

/**
 * Module dependencies.
 */

import Parser from './parser';
import Promise from 'bluebird';
import Requester from './requester';
import _ from 'lodash';
import debugnyan from 'debugnyan';
import methods from './methods';
import requestLogger from './logging/request-logger';
import semver from 'semver';

/**
 * Source arguments to find out if a callback has been passed.
 */

function source(...args) {
  const last = _.last(args);

  let callback;

  if (_.isFunction(last)) {
    callback = last;
    args = _.dropRight(args);
  }

  return [args, callback];
}

/**
 * List of networks and their default port mapping.
 */

const networks = {
  mainnet: 9333,
  testnet: 19333
};

/**
 * Constructor.
 */

class Client {
  constructor({
    agentOptions,
    headers = false,
    host = 'localhost',
    logger = debugnyan('vdinar-rpc'),
    network = 'mainnet',
    password,
    port,
    ssl = false,
    timeout = 30000,
    username,
    version
  } = {}) {
    if (!_.has(networks, network)) {
      throw new Error(`Invalid network name "${network}"`, { network });
    }

    this.agentOptions = agentOptions;
    this.auth = (password || username) && { pass: password, user: username };
    this.hasNamedParametersSupport = false;
    this.headers = headers;
    this.host = host;
    this.password = password;
    this.port = port || networks[network];
    this.ssl = {
      enabled: _.get(ssl, 'enabled', ssl),
      strict: _.get(ssl, 'strict', _.get(ssl, 'enabled', ssl))
    };
    this.timeout = timeout;

    // Version handling.
    if (version) {
      // Capture X.Y.Z when X.Y.Z.A is passed to support oddly formatted
      // versions such as 0.15.0.1.
      const result = /[0-9]+\.[0-9]+\.[0-9]+/.exec(version);

      if (!result) {
        throw new Error(`Invalid Version "${version}"`, { version });
      }

      [version] = result;

      this.hasNamedParametersSupport = semver.satisfies(version, '>=0.14.0');
    }

    this.version = version;
    this.methods = _.transform(methods, (result, method, name) => {
      result[_.toLower(name)] = {
        features: _.transform(method.features, (result, constraint, name) => {
          result[name] = {
            supported: version ? semver.satisfies(version, constraint) : true
          };
        }, {}),
        supported: version ? semver.satisfies(version, method.version) : true
      };
    }, {});

    const request = requestLogger(logger);

    this.request = Promise.promisifyAll(request.defaults({
      agentOptions: this.agentOptions,
      baseUrl: `${this.ssl.enabled ? 'https' : 'http'}://${this.host}:${this.port}`,
      strictSSL: this.ssl.strict,
      timeout: this.timeout
    }), { multiArgs: true });
    this.requester = new Requester({ methods: this.methods, version });
    this.parser = new Parser({ headers: this.headers });
  }

  /**
   * Execute `rpc` command.
   */

  command(...args) {
    let body;
    let callback;
    let [input, ...parameters] = args; // eslint-disable-line prefer-const
    const lastArg = _.last(parameters);
    const isBatch = Array.isArray(input);

    if (_.isFunction(lastArg)) {
      callback = lastArg;
      parameters = _.dropRight(parameters);
    }

    if (isBatch) {
      body = input.map((method, index) => this.requester.prepare({
        method: method.method,
        parameters: method.parameters,
        suffix: index
      }));
    } else {
      if (this.hasNamedParametersSupport && parameters.length === 1 && _.isPlainObject(parameters[0])) {
        parameters = parameters[0];
      }

      body = this.requester.prepare({ method: input, parameters });
    }

    return Promise.try(() => {
      return this.request.postAsync({
        auth: _.pickBy(this.auth, _.identity),
        body: JSON.stringify(body),
        uri: `/`
      }).bind(this)
        .then(this.parser.rpc);
    }).asCallback(callback);
  }
}

/**
 * Add all known RPC methods.
 */

_.forOwn(methods, (options, method) => {
  Client.prototype[method] = _.partial(Client.prototype.command, method.toLowerCase());
});

/**
 * Export Client class (ESM).
 */

export default Client;

/**
 * Export Client class (CJS) for compatibility with require('vdinar-rpc').
 */

module.exports = Client;
